bootstrap/kubeadm/api/v1alpha2/conversion.go

```diff
@@ -62,6 +75,14 @@ func (dst *KubeadmConfig) ConvertFrom(srcRaw conversion.Hub) error {
 		return err
 	}
 
+	for i, file := range src.Spec.Files {
+		if file.ContentFrom != nil {
+			// Cut file with referenced content.
+			// It's already in the marshaled data.
+			dst.Spec.Files = append(dst.Spec.Files[:i], dst.Spec.Files[i+1:]...)
+		}
+	}
```

Comment:
I'm not sure if this is advisable, if I'm reading the code here and above correctly, we're removing all the files with `file.ContentFrom` and then restoring it later?

Another thing we should probably do is to backport the changes we made on making `Content` non-required to v1alpah2

Reply:
We can index on path name and set contentFrom where content is unset. Unfortunately, content is a string, not a string pointer, so detecting unset vs. empty is not really possible otherwise.

The alternative is converting the secret content into raw exposed string, but that doesn't make much sense to me.

Thoughts on options?

Reply:
We can change content to string pointer, but I wasn't sure about that since it would require client changes

Reply:
I see, if we relax the `Content` field in v1alpha2, allowing it to be empty we should be able to set ContentFrom back regardless? It's probably fair to say that `ContentFrom` takes precedence over `Content`

Reply:
We could also start validating against empty string for content? But I could imagine sentinel files being useful for stuff, I don't really want to introduce that dependency.

Reply:
Whoops, racy comment timing. 

I think that makes sense. We'd only merge the contentFrom if a v1alpha3 file had the same path, and the v1alpha2 content for that file is empty.

In v1alpha3, either the content is non-empty string or contentFrom is populated, but in v1alpha2 we let content be empty?

Reply:
Today we don't, we'd have to update the CRDs to relax that requirement. Given that webhooks + CRDs should be deployed as a package (in lockstep), we should be fine

Reply:
Better?

Reply:
I didn't run into anything that seemed to cause issues with empty string in v1a2, let me know if that's somehow mistaken.

Reply:
I'm not sure if our tests would run into OpenAPI validation, but given that the v1alpha2 field is required https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/api/v1alpha2/kubeadmbootstrapconfig_types.go#L140 we need to relax that to allow v1alpha2 controllers to operate

Reply:
Isn’t an empty string valid? I thought that would mean the field would get serialized as empty string, and adding omitempty to string would omit the field?

Reply:
I think you're right, having it without omitempty makes an empty string a valid value